### PR TITLE
feat: create a routine for testing the nsfw model accuracy

### LINKED ISSUE #20: create a routine for testing the nsfw model accuracy
To evaluate the accuracy of our machine learning models, compare them, or estimate their efficiency, we need to test them with a provided database. In this task, we will create a routine to label the database and calculate the model accuracy.

### PR DESCRIPTION
**Pull request by MBSM110**
# Pull Request Checklist

Please check if your pull request fulfils the following requirements:

  * [x] The commit message follows our guidelines: <https://gitlab.playpod.ir/alpha/backend/nakama-module-next/-/blob/main/CONTRIBUTING.md#commit-message-format>
  * [x] Tests for the changes have been added / updated (for feat / fix / refactor)
  * [x] Docs for the changes have been added / updated (for feat / fix / refactor)

## Pull Request Type

What kind of change does this pull request introduce?

  * [ ] Build: Changes that affect the build system or external dependencies
  * [ ] CD: Changes to our CD configuration files and scripts
  * [ ] CI: Changes to our CI configuration files and scripts
  * [ ] Docs: Documentation only changes
  * [x] Feat: A new feature
  * [ ] Fix: A bug fix
  * [ ] Perf: A code change that improves performance
  * [ ] Refactor: A code change that neither fixes a bug nor adds a feature
  * [ ] Test: Adding missing tests or correcting existing tests

## What is the current behaviour?

Issue #20

## Does this Pull Request introduce a breaking change?

  * [ ] Yes
  * [x] No

## Other information

Closes #20

### CODE REVIEW

File: wolf/ml/w_utilities.hpp [hunk 1]
```diff
@@ -208,6 +214,44 @@ W_API auto get_env_string(_In_ const char *pKey) -> std::string;
...
+	the function stores the last processed image number in a separate history text
+	file, allowing the function to resume from where it left off if it is interrupted
+	or stopped before processing all images in the folder.
+
+	\param pImageFolderPath the image file path.
+	\param pLabeledImageTextFile the pLabeledImageTextFile file path.
+	\param pHistory the file that stores the last processed image number.
+	\return (void)
+*/
+W_API auto create_labeled_image_text(
```
Comment (PooyaEimandar) [inline wolf/ml/w_utilities.hpp:240]: using `auto` excessively can sometimes reduce code readability and may not always be appropriate, especially in cases where explicit type declarations are preferred for clarity. You are using `auto` along with defining return type. 

which one is better?
W_API std::vector<int> function() { return std::vector<int>(); }
or
W_API auto function()  -> std::vector<int> { return std::vector<int>(); }
or 
W_API auto function()  { return std::vector<int>();}

Comment (PooyaEimandar) [reply wolf/ml/w_utilities.hpp:240]: @MBSM110 @cthulhu-irl

Comment (cthulhu-irl) [reply wolf/ml/w_utilities.hpp:240]: @PooyaEimandar 

I would advise to use the following guideline:

first one is suited best for simple types like bool or w_status or other short name non-templated types,
second one for the templated/complex types and long names,
third one only when the type is unknown and its declaration would require complex syntax like `decltype`.

and if only one style must be followed/enforced, second one seems to be best suited.

Comment (cthulhu-irl) [reply wolf/ml/w_utilities.hpp:240]: @PooyaEimandar 

I would prefer second one. trailing return type is well suited for simple and short type names as well as complex (templated/etc) and long type names. and when the type is unknown or its declaration doesn't help with readability (like returning a combined ranges iterator) the third style matches it.

iirc it also can be enforced by clang-format/tidy.

File: wolf/ml/test/w_utilities_test.hpp [hunk 2]
```diff
@@ -304,13 +307,13 @@ BOOST_AUTO_TEST_CASE(get_env_cv_rect_function) {
...
 
-BOOST_AUTO_TEST_CASE(get_env_vector_of_int) {
+BOOST_AUTO_TEST_CASE(get_env_vector_of_int_function) {
   fs::path dot_env_file_path = utilities_asset_path / ".get_env_vector_of_int";
 
   set_env(dot_env_file_path.string().c_str());
 
   std::string key = "NUDITY_DETECTION_MODEL_PERMUTE";
-  std::vector<int> value = get_env_cv_rect(key.c_str());
+  std::vector<int> value = get_env_vector_of_int(key.c_str());
```
[the remark this answers is not in the document]
Comment (MBSM110) [reply wolf/ml/test/w_utilities_test.hpp:316]: The 'env' files present in the test directory are files that are stored within the project itself, making them an internal resource. These files maintain a constant state and are not dependent on any external resources.

File: wolf/ml/w_utilities.cpp [hunk 3]
```diff
@@ -450,6 +450,103 @@ auto get_env_vector_of_int(
...
+	_In_ const std::string pHistory) -> void
+{
+	std::vector<std::string> all_images_path = {};
+	if (std::filesystem::exists(pImageFolderPath))
+	{
+		all_images_path = images_in_directory(pImageFolderPath);
+	}
+	else
+	{
+		std::cout << "The path to images is not exist!!!" << std::endl;
```
[the remark this answers is not in the document]
Comment (MBSM110) [reply wolf/ml/w_utilities.cpp:517]: I would open a refactor task for modify this kind of "return error code".

File: wolf/ml/w_utilities.cpp [hunk 4]
```diff
@@ -450,6 +450,103 @@ auto get_env_vector_of_int(
...
 }
 
+auto map_key_value_to_label(
+	_In_ const int pKeyValue) -> int
+{
+	int label = 0;
+	switch (pKeyValue)
+	{
+		// 49 and 177 are keyboard key values related to "1"
+		case 49:
```
[the remark this answers is not in the document]
Comment (MBSM110) [reply wolf/ml/w_utilities.cpp:460]: I am not convinced that the way you suggest to change the code could resolve the issue you are considering.

[the remark this answers is not in the document]
Comment (cthulhu-irl) [reply wolf/ml/w_utilities.cpp:448]: self-explaining symbol names are durable, comments are not. for example typos in symbol names are caught by the compiler while comments can not be, being local to a specific function as the above code or shared throughout the codebase.

And as a general practice I'd also suggest to rather use comments to explain `why` and not the `what`, and embrace the language features and type system to properly address and/or convey meaning, usage, restrictions, and etc.

For this specific case, it's not vital (seemingly at least) but only suggested.

read more at here: https://refactoring.guru/replace-magic-number-with-symbolic-constant

File: wolf/ml/w_utilities.cpp [hunk 5]
```diff
@@ -450,6 +438,101 @@ auto get_env_vector_of_int(
...
+	{
+		image = cv::imread(all_images_path[i], cv::IMREAD_COLOR);
+		cv::imshow("The target image", image);
+		int pressed_key = cv::waitKey();
+		label = map_key_value_to_label(pressed_key);
+		std::string line_info = all_images_path[i] + " " + std::to_string(label);
+		write_in_file_append(
+			pLabeledImageTextFile,
+			line_info);
+	}
```
Comment (cthulhu-irl) [inline wolf/ml/w_utilities.cpp:531]: refactor to match wolf's style.

File: wolf/ml/w_utilities.cpp [hunk 6]
```diff
@@ -450,6 +438,101 @@ auto get_env_vector_of_int(
...
+	_In_ const std::string pLabeledImageTextFile,
+	_In_ const std::string pHistory) -> void
+{
+	std::vector<std::string> all_images_path = {};
+	if (std::filesystem::exists(pImageFolderPath))
+	{
+		all_images_path = images_in_directory(pImageFolderPath);
+	}
+	else
+	{
```
Comment (cthulhu-irl) [inline wolf/ml/w_utilities.cpp:504]: refactor to match wolf's style.

File: wolf/ml/w_utilities.cpp [hunk 7]
```diff
@@ -450,6 +438,101 @@ auto get_env_vector_of_int(
...
+	for (const auto& entry : fs::directory_iterator(pDirPath))
+	{
+		if (entry.is_regular_file() &&
+			(entry.path().extension() == ".jpeg" ||
+			 entry.path().extension() == ".jpg" ||
+			 entry.path().extension() == ".png"))
+		{
+			all_images_path.push_back(entry.path().string());
+		}
+	}
```
Comment (cthulhu-irl) [inline wolf/ml/w_utilities.cpp:488]: refactor to match wolf's style.

File: wolf/ml/w_utilities.cpp [hunk 8]
```diff
@@ -450,6 +438,101 @@ auto get_env_vector_of_int(
 	return int_vector;
 }
 
+auto map_key_value_to_label(
+	_In_ const int pKeyValue) -> int
+{
+	int label = 0;
+	switch (pKeyValue)
+	{
```
Comment (cthulhu-irl) [inline wolf/ml/w_utilities.cpp:446]: refactor to match wolf's style.

File: wolf/ml/nudity_detection/w_nudity_detection.hpp [hunk 9]
```diff
@@ -52,8 +48,16 @@ class w_nud_det {
   */
   W_API auto network_warm_up(_In_ int pHeight, _In_ int pWidth) -> void;
 
+  /*!
+  The function uses to calculate the accuracy of the input model over pre-labeled images. 
+  
+  	\param pInfoFilePath the path to the labeled file.
+  	\return the accuracy of model throw the given images
+  */
```
Comment (cthulhu-irl) [inline wolf/ml/nudity_detection/w_nudity_detection.hpp:56]: rewrite in wolf's documentation style: https://github.com/WolfEngine/wolf/blob/main/wolf/media/w_openal.hpp#L51-L56

File: wolf/ml/w_utilities.hpp [hunk 10]
```diff
@@ -223,6 +223,35 @@ W_API auto get_env_cv_rect(_In_ const char *pKey) -> cv::Rect;
...
+	used for further analysis or training of machine learning models. Additionally,
+	the function stores the last processed image number in a separate history text
+	file, allowing the function to resume from where it left off if it is interrupted
+	or stopped before processing all images in the folder.
+
+	\param pImageFolderPath the image file path.
+	\param pLabeledImageTextFile the pLabeledImageTextFile file path.
+	\param pHistory the file that stores the last processed image number.
+	\return (void)
+*/
```
Comment (cthulhu-irl) [inline wolf/ml/w_utilities.hpp:239]: rewrite in wolf's documentation style: https://github.com/WolfEngine/wolf/blob/main/wolf/media/w_openal.hpp#L51-L56

File: wolf/ml/w_utilities.hpp [hunk 11]
```diff
@@ -223,6 +223,35 @@ W_API auto get_env_cv_rect(_In_ const char *pKey) -> cv::Rect;
...
+	_In_ std::string pImageFolderPath,
+	_In_ std::string pLabeledImageTextFile,
+	_In_ std::string pHistory) -> void;
+
+/*!
+	The function return the value of an environment variable based on the input key.
+
+	\param pDirPath The path of image folder.
+	\return the result would be a vector contains image paths
+*/
```
Comment (cthulhu-irl) [inline wolf/ml/w_utilities.hpp:250]: rewrite in wolf's documentation style: https://github.com/WolfEngine/wolf/blob/main/wolf/media/w_openal.hpp#L51-L56

**Review by cthulhu-irl — CHANGES_REQUESTED** [review]

File: wolf/ml/w_utilities.cpp [hunk 12]
```diff
@@ -450,6 +438,101 @@ auto get_env_vector_of_int(
...
+
+	int n_images = all_images_path.size();
+
+	std::vector<std::string> text_info = {};
+	int last_processed_image_number = 0;
+	if (std::filesystem::exists(pHistory))
+	{
+		text_info = read_text_file_line_by_line(pHistory);
+		last_processed_image_number = std::stoi(text_info[0]);
+	}
```
[the remark this answers is not in the document]
Comment (MBSM110) [reply wolf/ml/w_utilities.cpp:516]: I should change the style in my whole folder. I will open a pull request for this later.

File: wolf/ml/w_utilities.hpp (hunk 10 above)
Comment (MBSM110) [reply wolf/ml/w_utilities.hpp:239]: I should change the style in my whole folder. I will open a pull request for this later.

**Review by cthulhu-irl — APPROVED** [review]
add issues for the following:
 - refactor ml code and documentation style
 - refactor to embrace error handling
 - refactor to embrace single responsibility principle (configuration and parameters instead of depending on env/file/etc)

**Pull request closed without merging**